### PR TITLE
Integration tests

### PR DESCRIPTION
**Pull request by srdjanns**
Spring Configuration classes for testing purposes, H2 database configuration, schema creation and data insert, integration tests.
All of work is done by Milica and Srdjan.

### CODE REVIEW

File: src/test/java/eu/execom/labs/test_integration_showcase/controller/PersonControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+                                                               .contentType(contentType))
+                                      .andReturn();
+        // then
+        assertEquals(mockResult.getResponse()
+                               .getStatus(),
+                400);
+    }
+
+    @Test
+    public void shouldAddPersonWhenSsnIsNotCorrect() throws Exception {
```
Comment (stanislavgrujic) [inline src/test/java/eu/execom/labs/test_integration_showcase/controller/PersonControllerTest.java]: Method name is incorrect -> `shouldReturnBadRequestWhenSsnIsInvalid`.

File: src/test/java/eu/execom/labs/test_integration_showcase/controller/PersonControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        MvcResult mockGetResult = mockMvc.perform(get("/persons/" + returnedPerson.getSsn()))
+                                         .andReturn();
+        String jsonGetResult = mockGetResult.getResponse()
+                                            .getContentAsString();
+        PersonDto resultPerson = objectMapper.readValue(jsonGetResult, PersonDto.class);
+        assertEquals(returnedPerson, resultPerson);
+    }
+
+    @Test
+    public void shouldGetOnePersonNotFound() throws Exception {
```
Comment (stanislavgrujic) [inline src/test/java/eu/execom/labs/test_integration_showcase/controller/PersonControllerTest.java]: Try to make method names as expressive as possible, e.g., `shouldReturnNotFoundOnMissingSsn`, or something similar.

**Comment by stanislavgrujic** [comment]
Please apply comments wherever applicable.

File: src/test/java/eu/execom/labs/test_integration_showcase/controller/PersonControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+                post("/persons").content(objectMapper.writeValueAsString(personDtoFirst))
+                                .contentType(contentType))
+                                           .andReturn();
+        MvcResult mockResultSecond = mockMvc.perform(
+                post("/persons").content(objectMapper.writeValueAsString(personDtoSecond))
+                                .contentType(contentType))
+                                            .andReturn();
+        assertEquals(mockResultFirst.getResponse()
+                                    .getStatus(),
+                200);
```
Comment (srdjanns) [inline src/test/java/eu/execom/labs/test_integration_showcase/controller/PersonControllerTest.java]: Execom formatter does all the job, it is his fault for this line organization. Should we turn it off or do something else about it?

Comment (stanislavgrujic) [inline src/test/java/eu/execom/labs/test_integration_showcase/controller/PersonControllerTest.java]: Try to make it work better? :) Don't spend too much time on it.

File: pom.xml [hunk 4]
```diff
@@ -10,15 +10,21 @@
...
         <validationapi.version>1.1.0.Final</validationapi.version>
-        <spring.data.version>1.10.2.RELEASE</spring.data.version>
         <mysql.version>5.1.6</mysql.version>
         <dbcp.version>1.4</dbcp.version>
+        <mockito.version>1.10.19</mockito.version>
+        <junit.version>4.12</junit.version>
+        <dbh2.version>1.4.187</dbh2.version>
         <lombok.version>1.16.10</lombok.version>
         <jackson.version>2.8.1</jackson.version>
+        <jackson.mapper.asl.version>1.9.13</jackson.mapper.asl.version>
```
Comment (stanislavgrujic) [inline pom.xml]: You should not be mixing two different versions of jackson - please use `jackson-databind` instead, which should have a matching `jackson.version` as above. Also, the latest version is 2.8.5

File: pom.xml [hunk 5]
```diff
@@ -92,7 +97,83 @@
...
+        <dependency>
+            <groupId>com.jayway.jsonpath</groupId>
+            <artifactId>json-path</artifactId>
+            <version>${json.path.version}</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.json</groupId>
+            <artifactId>json</artifactId>
+            <version>20090211</version>
```
Comment (stanislavgrujic) [inline pom.xml]: This is very old version, there should be a newer one - 20160810. Also, I would say that you only use it in test scope, so that should be used.

File: src/test/java/eu/execom/labs/test_integration_showcase/configuration/DataConfigTest.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+import org.springframework.orm.jpa.JpaTransactionManager;
+import org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean;
+
+import eu.execom.labs.test_integration_showcase.entity.EntityComponents;
+import eu.execom.labs.test_integration_showcase.repository.RepositoryComponents;
+
+@Configuration
+@EnableJpaRepositories(basePackageClasses = RepositoryComponents.class)
+@PropertySource({"classpath:jdbc.test.properties"})
+public class DataConfigTest {
```
Comment (stanislavgrujic) [inline src/test/java/eu/execom/labs/test_integration_showcase/configuration/DataConfigTest.java]: Class name implies that this is a test for DataConfig, which is not true. Maybe would be better to rename it to `TestConfiguration` or something similar.

File: src/test/java/eu/execom/labs/test_integration_showcase/configuration/RootConfigTest.java [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+import org.springframework.context.annotation.ComponentScan;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Import;
+
+import eu.execom.labs.test_integration_showcase.service.ServiceComponents;
+
+@Configuration
+@Import({DataConfigTest.class})
+@ComponentScan(basePackageClasses = ServiceComponents.class)
+public class RootConfigTest {
```
Comment (stanislavgrujic) [inline src/test/java/eu/execom/labs/test_integration_showcase/configuration/RootConfigTest.java]: Same as above, please rename it to `TestRootConfiguration`, or similar.

File: src/test/java/eu/execom/labs/test_integration_showcase/configuration/WebConfigTest.java [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+import org.springframework.context.annotation.Configuration;
+import org.springframework.web.servlet.config.annotation.EnableWebMvc;
+import org.springframework.web.servlet.config.annotation.WebMvcConfigurerAdapter;
+
+import eu.execom.labs.test_integration_showcase.controller.ControllerComponents;
+
+@Configuration
+@EnableWebMvc
+@ComponentScan(basePackageClasses = ControllerComponents.class)
+public class WebConfigTest extends WebMvcConfigurerAdapter {
```
Comment (stanislavgrujic) [inline src/test/java/eu/execom/labs/test_integration_showcase/configuration/WebConfigTest.java]: See my comment above.

File: src/test/java/eu/execom/labs/test_integration_showcase/controller/PersonControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+    private MediaType contentType = MediaType.APPLICATION_JSON;
+
+    @Before
+    public void setUp() {
+        mockMvc = MockMvcBuilders.webAppContextSetup(webApplicationContext)
+                                 .build();
+        objectMapper = new ObjectMapper();
+    }
+
+    public PersonDto createPerson(String email, String ssn) {
```
Comment (stanislavgrujic) [inline src/test/java/eu/execom/labs/test_integration_showcase/controller/PersonControllerTest.java]: This method should be private, and as such it should go below all places where it is used, not above all of them.

**Comment by stanislavgrujic** [comment]
Please add updated code formatter configuration to the source control.

File: src/test/java/eu/execom/labs/test_integration_showcase/controller/PersonControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+public class PersonControllerTest {
+
+    @Autowired
+    private WebApplicationContext webApplicationContext;
+
+    private MockMvc mockMvc;
+
+    private ObjectMapper objectMapper;
+
+    private MediaType contentType = MediaType.APPLICATION_JSON;
```
Comment (stanislavgrujic) [inline src/test/java/eu/execom/labs/test_integration_showcase/controller/PersonControllerTest.java]: Please remove this field where it is used, and use constant directly.

File: src/test/java/eu/execom/labs/test_integration_showcase/controller/PersonControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+
+    private PersonDto createPerson(String email, String ssn) {
+        PersonDto personDto = new PersonDto();
+        personDto.setEmail(email);
+        personDto.setFirstName("John");
+        personDto.setLastName("Doe");
+        personDto.setSsn(ssn);
+        return personDto;
+    }
+}
```
Comment (stanislavgrujic) [inline src/test/java/eu/execom/labs/test_integration_showcase/controller/PersonControllerTest.java]: What happens if you try to insert duplicate SSN? I don't see a test case for that.

Comment (mpoznanovic) [inline src/test/java/eu/execom/labs/test_integration_showcase/controller/PersonControllerTest.java]: We have exception handler for duplicate entry. Also, there is a test case for adding duplicate person.

Comment (stanislavgrujic) [inline src/test/java/eu/execom/labs/test_integration_showcase/controller/PersonControllerTest.java]: And how do you know, based on which check the application fails - duplicate e-mail, or duplicate ssn? It would be good to have this test case split into two, to make sure both checks are equally tested.
